ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -88,89 +91,104 @@ socketAsMuxBearer sd = do
-    bracketOnError
-        (socket (addrFamily addr) Stream defaultProtocol)
-        close
-        (\sd -> do
-            setSocketOption sd ReuseAddr 1
-            setSocketOption sd ReusePort 1
-            bind sd (addrAddress addr)
-            listen sd 2
-            rh <- async (server sd)
-            return (sd, rh)
-        )
-  where
-    server sd = forever $ do
-        (client, _) <- accept sd
-        aid <- async $ larval client
-        void $ async $ watcher client aid
-
-    larval sd = do
-        bearer <- socketAsMuxBearer sd
-        Mx.muxBearerSetState bearer Mx.Connected
-        Mx.muxStart verMpds bearer Mx.StyleServer rescb_m
-
-    watcher sd aid = do
-        res_e <- waitCatch aid
-        case rescb_m of
-             Nothing ->
-                 case res_e of
-                      Left  e -> close sd >> say ("Version negotiation failed with " ++ show e)
-                      Right _ -> return ()
-             Just rescb ->
-                 case res_e of
-                      Left e  -> do
-                          rescb $ Just e
-                          close sd
-                      Right _ -> rescb Nothing
-
-killResponder :: (Socket, Async ()) -> IO ()
-killResponder (sd, hdl) = do
-    cancel hdl
-    close sd
-
-startInitiator :: (Mx.ProtocolEnum ptcl, Ord ptcl, Enum ptcl, Bounded ptcl)
-               => [(Mx.Version, Mx.MiniProtocolDescriptions ptcl IO)]
-               -> AddrInfo
-               -> AddrInfo
-               -> IO ()
-startInitiator mpds local remote = startInitiatorT mpds local remote Nothing
-
-startInitiatorT :: (Mx.ProtocolEnum ptcl, Ord ptcl, Enum ptcl, Bounded ptcl)
-                => [(Mx.Version, Mx.MiniProtocolDescriptions ptcl IO)]
-                -> AddrInfo
-                -> AddrInfo
-                -> Maybe (Maybe SomeException -> IO ())
-                -> IO ()
-startInitiatorT verMpds local remote rescb_m =
-    bracketOnError
-        (socket (addrFamily local) Stream defaultProtocol)
-        close
-        (\sd -> do
-            bearer <- socketAsMuxBearer sd
-            setSocketOption sd ReuseAddr 1
-            setSocketOption sd ReusePort 1
-            bind sd (addrAddress local)
-            connect sd (addrAddress remote)
-            Mx.muxBearerSetState bearer Mx.Connected
-
-            void $ Mx.muxStart verMpds bearer Mx.StyleClient rescb_m
-        )
-
 hexDump :: BL.ByteString -> String -> IO ()
 hexDump buf out | BL.empty == buf = say out
 hexDump buf out = hexDump (BL.tail buf) (out ++ printf "0x%02x " (BL.head buf))
 
+-- |
+-- Like @'runNetworkNodeWithSocket'@ but it allows to run an action when an
+-- exception is raised by the mux layer.  This is useful in tests.
+-- 
+runNetworkNodeWithSocket'
+  :: forall ptcl.
+     ( Mx.ProtocolEnum ptcl
+     , Ord ptcl
+     , Enum ptcl
+     , Bounded ptcl
+     )
+  => NetworkInterface ptcl AddrInfo IO
+  -> Maybe (Maybe SomeException -> IO ())
+  -> IO (NetworkNode AddrInfo IO)
```

Comment:
Hmm. It's not immediately obvious that `NetworkInterface` is a useful abstraction. Yes it looks like `NetworkNode` is. But it looks like there's many ways to make a `NetworkNode`, test, sockets, pipes etc. But is `NetworkInterface` just some arguments to some of the functions that construct a `NetworkNode`? Does it get used elsewhere that I missed?

Reply:
@duncan the idea here is to gather all the necessary arguments in one place.  If one wants to get `NetworkNode` one needs to present `NetworkInterface`. The second argument will eventually dissapear when `iohk-monitoring-network` will get included and `ouroboros-network` will get proper loging, that's why I don't count it.

This works now only for sockets, and I hope this will also be useful for the dynamic-pipes for which @karknu would like to give the same interface as for sockets (metioned in a [slack](https://input-output-rnd.slack.com/archives/GFLJKNG3T/p1552484282369900) conversation).  For bare `pipes` it's not needed, but for bare pipes we also don't need `NetworkNode`.